File: src/nds/Picross/py/GhidraArm9Export.java
// Ghidra headless postScript: ARM9 function-level export (Java, pure ASCII)
// Usage: analyzeHeadless <proj> <name> -process arm9_full.bin -scriptPath <dir> -postScript GhidraArm9Export.java
// Output: <scriptPath>/arm9_export/funcs_flat.json  callgraph.json  funcs_asm/
import java.io.File;
import java.io.FileWriter;
import java.util.*;

import ghidra.app.script.GhidraScript;
import ghidra.program.model.address.Address;
import ghidra.program.model.listing.*;
import ghidra.program.model.mem.Memory;
import ghidra.program.model.mem.MemoryBlock;
import ghidra.program.model.symbol.*;

public class GhidraArm9Export extends GhidraScript {

    private static final int MAX_DEPTH = 16;

    private final Set<Long> visited = new HashSet<Long>();
    private final Map<String, Map<String, Object>> callGraph = new LinkedHashMap<String, Map<String, Object>>();
    private File asmDir;
    private FunctionManager fm;

    @Override
    public void run() throws Exception {
        // ---- 0. image info ----
        long base = currentProgram.getImageBase().getOffset();
        println("IMAGE BASE: 0x" + Long.toHexString(base));
        Memory mem = currentProgram.getMemory();
        for (MemoryBlock b : mem.getBlocks()) {
            println("BLOCK " + b.getName() + " " + hex(b.getStart().getOffset()) + "-" + hex(b.getEnd().getOffset()) + " size=0x" + Long.toHexString(b.getSize()));
        }

        // ARM9 main entry: 0x02003000 if base is high, else 0x3000 (file-offset import)
        long rootAddr = (base < 0x10000L) ? 0x00003000L : 0x02003000L;

        File outRoot = new File(getSourceFile().getParentFile(), "arm9_export");
        asmDir = new File(outRoot, "funcs_asm");
        if (!asmDir.exists()) asmDir.mkdirs();

        fm = getFunctionManager();

        // ---- 1. flat table ----
        List<Map<String, Object>> flat = new ArrayList<Map<String, Object>>();
        for (Function f : fm.getFunctions(true)) {
            long ep = f.getEntryPoint().getOffset();
            Address min = f.getBody().getMinAddress();
            Address max = f.getBody().getMaxAddress();
            Map<String, Object> m = new LinkedHashMap<String, Object>();
            m.put("entry", hex(ep & ~1L));
            m.put("entry_raw", hex(ep));
            m.put("start", hex(min.getOffset()));
            m.put("end", hex(max.getOffset()));
            m.put("size", max.getOffset() - min.getOffset() + 1);
            m.put("name", f.getName());
            m.put("is_thunk", Boolean.valueOf(f.isThunk()));
            m.put("is_thumb", Boolean.valueOf((ep & 1) == 1));
            flat.add(m);
        }
        Collections.sort(flat, new Comparator<Map<String, Object>>() {
            public int compare(Map<String, Object> a, Map<String, Object> b) {
                return ((String) a.get("entry")).compareTo((String) b.get("entry"));
            }
        });
        writeJson(new File(outRoot, "funcs_flat.json"), flat);
        println("FLAT count = " + flat.size());

        // ---- 2. DFS from main entry ----
        Function rootFunc = fm.getFunctionAt(toAddr(rootAddr));
        if (rootFunc == null) rootFunc = fm.getFunctionContaining(toAddr(rootAddr));
        if (rootFunc == null) {
            println("ERROR: root func not found @ " + hex(rootAddr));
            return;
        }
        dfs(rootFunc, 0);
        writeJson(new File(outRoot, "callgraph.json"), callGraph);
        println("EXPORT done: flat=" + flat.size() + " dfs=" + visited.size());
    }

    private void dfs(Function f, int depth) throws Exception {
        long entry = f.getEntryPoint().getOffset() & ~1L;
        if (visited.contains(entry) || depth > MAX_DEPTH) return;
        visited.add(entry);
        String hexE = hex(entry);

        StringBuilder sb = new StringBuilder();
        CodeUnitIterator cus = getListing().getCodeUnits(f.getBody(), true);
        while (cus.hasNext()) {
            CodeUnit cu = cus.next();
            sb.append(String.format("0x%08X\t%s%n", cu.getAddress().getOffset(), cu.toString()));
        }

        Set<String> callees = new TreeSet<String>();
        ReferenceIterator refs = getReferenceManager().getReferencesFrom(f.getBody());
        while (refs.hasNext()) {
            Reference r = refs.next();
            RefType ft = r.getFlowType();
            if (ft.equals(RefType.CALL) || ft.equals(RefType.UNCONDITIONAL_JUMP)) {
                Function tf = fm.getFunctionContaining(r.getToAddress());
                if (tf != null) callees.add(hex(tf.getEntryPoint().getOffset() & ~1L));
            }
        }

        Map<String, Object> meta = new LinkedHashMap<String, Object>();
        meta.put("func_name", f.getName());
        meta.put("entry", hexE);
        meta.put("start", hex(f.getBody().getMinAddress().getOffset()));
        meta.put("end", hex(f.getBody().getMaxAddress().getOffset()));
        meta.put("depth", Integer.valueOf(depth));
        meta.put("is_thumb", Boolean.valueOf((f.getEntryPoint().getOffset() & 1) == 1));
        meta.put("called_subfuncs", new ArrayList<String>(callees));
        callGraph.put(hexE, meta);

        FileWriter w = new FileWriter(new File(asmDir, safe(hexE + "_" + f.getName()) + ".asm"));
        w.write("FUNC_META: " + json(meta) + "\n\n=====DISASM=====\n" + sb.toString());
        w.close();

        for (String h : callees) {
            Function sub = fm.getFunctionAt(toAddr(h));
            if (sub == null) sub = fm.getFunctionContaining(toAddr(h));
            if (sub != null) dfs(sub, depth + 1);
        }
    }

    private Address toAddr(long v) {
        return currentProgram.getAddressFactory().getDefaultAddressSpace().getAddress(v);
    }

    private static String hex(long v) {
        return String.format("0x%08X", v);
    }

    private static String safe(String s) {
        return s.replaceAll("[\\\\/:*?\"<>|]", "_");
    }

    private void writeJson(File f, Object o) throws Exception {
        FileWriter w = new FileWriter(f);
        w.write(json(o));
        w.close();
    }

    private static String json(Object o) {
        if (o instanceof Map) {
            StringBuilder sb = new StringBuilder("{");
            boolean first = true;
            for (Map.Entry<?, ?> e : ((Map<?, ?>) o).entrySet()) {
                if (!first) sb.append(",");
                first = false;
                sb.append("\"").append(e.getKey()).append("\":").append(json(e.getValue()));
            }
            return sb.append("}").toString();
        }
        if (o instanceof List) {
            StringBuilder sb = new StringBuilder("[");
            boolean first = true;
            for (Object x : (List<?>) o) {
                if (!first) sb.append(",");
                first = false;
                sb.append(json(x));
            }
            return sb.append("]").toString();
        }
        if (o instanceof String) return "\"" + o + "\"";
        if (o instanceof Boolean || o instanceof Number) return String.valueOf(o);
        return "\"" + String.valueOf(o) + "\"";
    }
}

File: src/nds/Picross/py/ghidra_arm9_export.py
#@category Analysis
# Ghidra headless postScript: ARM9 function-level export (PyGhidra / Python3)
# Output: <scriptPath>/arm9_export/funcs_flat.json  callgraph.json  funcs_asm/
import os
import json
import re

from ghidra.program.model.symbol import RefType

ROOT = 0x02003000
MAX_DEPTH = 16


def hexl(v):
    return "0x%08X" % (int(v) & 0xFFFFFFFF)


def safe(s):
    return re.sub(r'[\\/:*?"<>|]', "_", str(s))


def to_addr(v):
    return currentProgram.getAddressFactory().getDefaultAddressSpace().getAddress(int(v))


# ---- 0. image info (verify base) ----
base = currentProgram.getImageBase().getOffset()
print("IMAGE BASE: 0x%08X" % base)
for b in currentProgram.getMemory().getBlocks():
    print("BLOCK %s %s-%s size=0x%x" % (b.getName(), hexl(b.getStart().getOffset()),
                                        hexl(b.getEnd().getOffset()), b.getSize()))

# ARM9 main entry: high-base or file-offset import
root_addr = 0x00003000 if base < 0x10000 else ROOT

fm = currentProgram.getFunctionManager()
listing = currentProgram.getListing()
ref_mgr = currentProgram.getReferenceManager()

out_root = os.path.join(os.path.dirname(getSourceFile().getAbsolutePath()), "arm9_export")
asm_dir = os.path.join(out_root, "funcs_asm")
if not os.path.exists(asm_dir):
    os.makedirs(asm_dir)

# ---- 1. flat table ----
flat = []
it = fm.getFunctions(True)
while it.hasNext():
    f = it.next()
    ep = f.getEntryPoint().getOffset()
    min_a = f.getBody().getMinAddress().getOffset()
    max_a = f.getBody().getMaxAddress().getOffset()
    flat.append({
        "entry": hexl(ep & ~1),
        "entry_raw": hexl(ep),
        "start": hexl(min_a),
        "end": hexl(max_a),
        "size": max_a - min_a + 1,
        "name": f.getName(),
        "is_thunk": f.isThunk(),
        "is_thumb": (ep & 1) == 1,
    })
flat.sort(key=lambda m: m["entry"])
with open(os.path.join(out_root, "funcs_flat.json"), "w") as w:
    w.write(json.dumps(flat, indent=1))
print("FLAT count = %d" % len(flat))

# ---- 2. DFS from main entry ----
visited = set()
callgraph = {}


def dfs(f, depth):
    if f is None:
        return
    entry = f.getEntryPoint().getOffset() & ~1
    if entry in visited or depth > MAX_DEPTH:
        return
    visited.add(entry)
    he = hexl(entry)

    sb = []
    cus = listing.getCodeUnits(f.getBody(), True)
    while cus.hasNext():
        cu = cus.next()
        sb.append("0x%08X\t%s" % (cu.getAddress().getOffset(), cu.toString()))

    callees = set()
    refs = ref_mgr.getReferencesFrom(f.getBody())
    while refs.hasNext():
        r = refs.next()
        ft = r.getFlowType()
        if ft.equals(RefType.CALL) or ft.equals(RefType.UNCONDITIONAL_JUMP):
            tf = fm.getFunctionContaining(r.getToAddress())
            if tf is not None:
                callees.add(hexl(tf.getEntryPoint().getOffset() & ~1))

    meta = {
        "func_name": f.getName(),
        "entry": he,
        "start": hexl(f.getBody().getMinAddress().getOffset()),
        "end": hexl(f.getBody().getMaxAddress().getOffset()),
        "depth": depth,
        "is_thumb": (f.getEntryPoint().getOffset() & 1) == 1,
        "called_subfuncs": sorted(callees),
    }
    callgraph[he] = meta

    with open(os.path.join(asm_dir, safe(he + "_" + f.getName()) + ".asm"), "w") as w:
        w.write("FUNC_META: " + json.dumps(meta) + "\n\n=====DISASM=====\n" + "\n".join(sb))

    for h in sorted(callees):
        sub = fm.getFunctionAt(to_addr(h))
        if sub is None:
            sub = fm.getFunctionContaining(to_addr(h))
        if sub is not None:
            dfs(sub, depth + 1)


root_f = fm.getFunctionAt(to_addr(root_addr))
if root_f is None:
    root_f = fm.getFunctionContaining(to_addr(root_addr))
if root_f is None:
    print("ERROR: root func not found @ " + hexl(root_addr))
else:
    dfs(root_f, 0)
    with open(os.path.join(out_root, "callgraph.json"), "w") as w:
        w.write(json.dumps(callgraph, indent=1))
    print("EXPORT done: flat=%d dfs=%d" % (len(flat), len(visited)))
